app/controllers/assignments_controller.rb

```diff
@@ -101,82 +101,203 @@ def create
+      if !(isCodeFileUploaded)
+        # create asssignment but don't process it
+        redirect_to course_assignments_url(@course), notice: 'Assignment was successfully created. Please upload the student submission files to preview files and/or process the assignment'
+    
+      # ascertain that all the fields are valid
+      else
+        # verify codeFile and mapFile (if uploaded)
+        isValidCodeFile = is_valid_zip?(codeFile.content_type, codeFile.path)
+        isValidMapFile = is_valid_map_or_no_map?(isOptionalOptionsEnabled, mapFile)
 
-      # No student submission file was uploaded
-      if params[:assignment]["file"].nil?
-        # Create asssignment but don't process it
-        redirect_to course_assignments_url(@course), notice: 'Assignment was successfully created. Please upload the student submission files and mapping file (if any) to process the assignment'
-      # Student submission file is a valid zip
-      elsif (is_valid_zip?(params[:assignment]["file"].content_type, params[:assignment]["file"].path))
-        # Don't process the file and show error if the mapping was enabled but no mapping file was uploaded
-        if (is_valid_map_or_no_map?(isMapEnabled, params[:assignment]["mapfile"])) 
-          self.start_upload(@assignment, params[:assignment]["file"], isMapEnabled, params[:assignment]["mapfile"])
-          redirect_to course_assignments_url(@course), notice: 'Assignment was successfully created. Please refresh this page after a few minutes to view the similarity results.'
-        # Don't process the file and show error if the mapping was enabled but no mapping file was uploaded
-        elsif (isMapEnabled && params[:assignment]["mapfile"].nil?)
-          @assignment.errors.add(:mapfile, "containing mapped student names need to be uploaded if the 'Upload map file' box is ticked")
-          return render action: "new"
-        # Don't process the file and show error if the mapping was enabled but no mapping file was uploaded
-        else 
-          @assignment.errors.add :mapfile, "containing mapped student names must be a valid csv file"
-          return render action: "new"
+        # uploaded student code file was not valid
+        if (!isValidCodeFile)
+          @assignment.errors.add :file, "Containing student submission files must be a valid zip file"
+          hasErrorParams = true
         end
-      # Student submission file is not a valid zip file
-      else
-        @assignment.errors.add :file, "containing student submission files must be a valid zip file"
-        if (params[:assignment]["mapfile"].nil? && isMapEnabled)
-          @assignment.errors.add :mapfile, "containing mapped student names need to be uploaded if the 'Upload map file' box is ticked"
-        elsif !(is_valid_map_or_no_map?(isMapEnabled, params[:assignment]["mapfile"])) 
-          @assignment.errors.add :mapfile, "containing mapped student names must be a valid csv file"
+        
```

Comment:
In MVC/Rails, people usually follow "Fat models, skinny controllers" convention for code's readability and maintenance. Fat model is not always maintainable. So, people introduce service layer to do validation or to interact with models. 

In this case, it's suggested to have assignment validation service for code segments: is_valid_zip?, is_valid_map_or_no_map? ... and to have a file helper service for code segment: extract_uploaded_files ...
It will help to make the code low coupling, easy to read and maintain.

Reply:
@knmnyn do you have any comments on it?